### PR TITLE
Encapsulate the timbot logic into a Timbot class.

### PR DESCRIPTION
**Pull request by timflannagan**

### CODE REVIEW

**Review by nathan-weinberg — CHANGES_REQUESTED** [review]
LGTM; can you link the test results?

**Comment by timflannagan** [comment]
@nathan-weinberg Test results were as followed:
```
Timbot: Passing
```

**Comment by timflannagan** [comment]
I think it's safe to say we can merge this :+1: :tada:

File: lib/timbot.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+		self.send_message('I\'m in. Best 2 games out of 3?')
+
+	def default_message_handler(self):
+		self.send_message('keep pounding')
+
+	def chose_message_handler(self, message):
+		return self.message_handler[next(key for key in self.message_handler if key in message)]
+
+	def get_channel_history(self):
+		return self.client.api_call("channel.history", channel=self.channel_id, count=1)
```
Comment (Daniel-Pivonka) [inline lib/timbot.py]: we need 'group.history' not 'channel.history'

**Review by Daniel-Pivonka — CHANGES_REQUESTED** [review]

File: scripts/run-timbot-local.sh [hunk 2]
```diff
@@ -14,12 +14,14 @@ elif [[ -z "$SLACK_TIMBOT_USER_ID" ]]; then
 fi
 
 CONTAINER_RUN_CMD=${CONTAINER_RUN_CMD:-"podman run"}
+CONTAIMER_IMAGE_NAME=${CONTAINER_IMAGE_NAME:-"timbot-local"}
```
Comment (Daniel-Pivonka) [inline scripts/run-timbot-local.sh]: miss spelling of CONTAINER

File: lib/timbot.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+			history = self.get_channel_history()
+
+			if datetime.datetime.now().strftime('%H:%M') == self.ideal_lunch_time:
+				self.send_lunch_time_notification()
+
+			# wait until there are messages populated in the history dictionary
+			if 'messages' not in history or len(history['messages']) == 0:
+				time.sleep(1.5)
+
+			for data in history['messages']:
```
Comment (Daniel-Pivonka) [inline lib/timbot.py:115]: key 'messages' is not checked

File: lib/timbot.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+		while True:
+			history = self.get_channel_history()
+
+			if datetime.datetime.now().strftime('%H:%M') == self.ideal_lunch_time:
+				self.send_lunch_time_notification()
+
+			# wait until there are messages populated in the history dictionary
+			if 'messages' not in history or len(history['messages']) == 0:
+				time.sleep(1.5)
```
Comment (Daniel-Pivonka) [inline lib/timbot.py:112]: this logic will not work. you get the history -> if the key 'messages' is not present u sleep -> you processed to use the key on line 114.

File: lib/timbot.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+		self.send_message('keep pounding')
+
+	def chose_message_handler(self, message):
+		return self.message_handler[next(key for key in self.message_handler if key in message)]
+
+	def get_channel_history(self):
+		return self.client.api_call("channel.history", channel=self.channel_id, count=1)
+
+	def send_message(self, text):
+		self.client.api_call('chat.postMessage', channel=self.channel_id, text=text, as_user=False)
```
Comment (Daniel-Pivonka) [inline lib/timbot.py:42]: messages need to be sent as the user

File: lib/helpers.py [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+def utc_to_est(utc_time):
+	return str((int(utc_time[:2]) - 5) % 24) + ':' + utc_time[-2:]
+
+def choose_lunchcation():
+	places = ['pauls', 'asian plus', 'moes', 'the 99', 'chilis']
+	weights = [0.5, 0.3, 0.1, 0.05, 0.05]
+	choice = np.random.choice(places, p=weights)
+	return choice
+
+def print_json(json_dict):
```
Comment (Daniel-Pivonka) [inline lib/helpers.py]: where is this used?

File: lib/timbot.py [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+			return
+
+		if 'where' in message:
+			if (datetime.datetime.today().weekday() == self.weekday_lunch_index) or 'no feast' in message:
+				random.seed(datetime.datetime.now())
+				self.send_message(helpers.choose_lunchcation())
+			else:
+				self.send_message(self.default_lunchcation)
+		elif 'when' in message or 'time' in message:
+			self.send_chat_image("lunchcharts.png", '', 'IdealLunchTimeChart', self.ideal_lunch_time)
```
Comment (Daniel-Pivonka) [inline lib/timbot.py]: why was this 'send_message(utc_to_est(ideal_lunch_time))' removed?

Comment (Daniel-Pivonka) [inline lib/timbot.py]: the image is 'lunchchart.png'

**Comment by Daniel-Pivonka** [comment]
This is a big change to the structure of the project. I've found lots of errors and i didn't even test all the features. Please fix the errors I found and test all of the features for this pr to be considered

File: lib/timbot.py (hunk 4 above)
Comment (timflannagan) [reply lib/timbot.py:112]: lol whoops

File: lib/timbot.py (hunk 3 above)
Comment (timflannagan) [reply lib/timbot.py:115]: It wouldn't need to be checked as we verified there's a message key in the `while` statement above.

**Comment by timflannagan** [comment]
@Daniel-Pivonka  @nathan-weinberg It looks like this can be merged! :tada: 
Good work everyone!

**Comment by Daniel-Pivonka** [comment]
@timflannagan1 please provide your test logs

**Comment by timflannagan** [comment]
@Daniel-Pivonka Absolutely, I'll paste my most recent run below:
```bash
$ ./scripts/test-timbot.sh
VERIFIED timbot works
```

**Comment by timflannagan** [comment]
Oh, weird you must have accidentally closed my PR! That's alright, I'll re-open it.

**Comment by Daniel-Pivonka** [comment]
merge conflicts and invalid test results closing

**Comment by nathan-weinberg** [comment]
sad

**Comment by nathan-weinberg** [comment]
this PR is officially in development hell

**Pull request closed without merging**